src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -505,55 +530,90 @@ private void onMqttSubscribe(
                     }
                 }
 
-                for (int progress = offset; progress < limit; progress = subscription.limit())
+                for (int progress = offset, index = 0; progress < limit; progress = subscription.limit(), index++)
                 {
                     subscription = mqttSubscriptionRO.tryWrap(buffer, progress, limit);
                     if (subscription == null)
                     {
                         break;
                     }
 
-                    int reasonCode = 0x8f;
+                    // TODO - order of reason codes MATTER. first stream -> first entry into list, etc
+                    //      - need to be able to map the correct reason codes to the correct streams
+                    //      - defer reason code assignment to WINDOW
+                    // need to map reason code -> stream in order of when they were created, not when they come in
+                    // 	- ie. streams: 0, 1, 2, 3
+                    //
+                    // 		even if 1 sends WINDOW before 0, should sitll map to list[1], not list[0]
+                    //      would streams be assigned a number to indicate their reasoncodesIndex?
+                    // int reasonCode = 0x8f;  // 0x8F - Topic Filter invalid
                     final String topicFilter = subscription.topicFilter().asString();
                     final RouteFW route = resolveTarget(routeId, authorization, topicFilter);
 
                     if (route != null)
                     {
-                        final MqttServerStream serverStream = new MqttServerStream();
-
                         final long newRouteId = route.correlationId();
                         final long newInitialId = supplyInitialId.applyAsLong(newRouteId);
                         final long newReplyId = supplyReplyId.applyAsLong(newInitialId);
 
                         final MessageConsumer newTarget = router.supplyReceiver(newInitialId);
 
+                        final MqttServerStream serverStream = new MqttServerStream(this, newTarget,
+                            newRouteId, newInitialId, newReplyId, index);
+
                         final MqttBeginExFW beginEx = mqttBeginExRW
-                            .wrap(writeBuffer, BeginFW.FIELD_OFFSET_EXTENSION, writeBuffer.capacity())
+                            .wrap(extBuffer, 0, extBuffer.capacity())
                             .typeId(mqttTypeId)
                             .role(r -> r.set(MqttRole.RECEIVER))
                             .clientId("client")
                             .topic(topicFilter)
                             .subscriptionId(subscriptionId)
                             .build();
 
-                        serverStream.doMqttBeginEx(newTarget, newRouteId, newInitialId,
-                            decodeTraceId, beginEx.buffer(), beginEx.offset(), beginEx.sizeof());
+                        serverStream.doMqttBeginEx(decodeTraceId, beginEx);
```

Comment:
Suggest moving the `MqttBeginEx` construction into the `serverStream.doMqttBegin(...)` instead.